Time score added

WendyFu004 opened the pull request:
Now the final score would included the time score and Final score = accuracyScore - timeScore
It was sum before I changed to minus here since time score is the time spent so far. The accuracy score is very huge now, we can balance the weight or find a new time score rule.
This branch will works on the time score only. The TrackImplement branch won't have any update about this part.

WendyFu004 committed:
Time score added

Now the final score would included the time score and Final score = accuracyScore - timeScore
It was sum before I changed to minus here since time score is the time spent so far. The accuracy score is very huge now, we can balance the weight or find a new time score rule.
This branch will works on the time score only. The TrackImplement branch won't have any update about this part.

Aether700 reviewed (changes requested):
Would it be possible to have the time score be a "positive" number? as in the greater the time score the greater the final score?

For example we could have the first racer to finish always having a set maximum score and then decrease the score of the other racers based on the time they take compared to that initial time?

WendyFu004 committed:
positive time score

The bullseye will recode the time fist racer hit it and the time score will lerp between max and min given by player in the scoremanager script.

Aether700 commented on Assets/Scripts/Bullseye.cs:
```diff
@@ -21,4 +23,16 @@ public bool HasCollidedWith(Vector3 pos)
         return Vector3.Distance(pos, transform.position) < (transform.localScale.x / 2);
     }
+
+    private void OnTriggerEnter(Collider other)
```
please do not use OnTriggerEnter inside the bullseye target as it is technically possible to collide with the bullseye without hitting it's graphics given that it uses a box collider (there are no cylinder collider in unity)

Instead you can fetch the time of the first finished racer in the racer manager by checking if it's the first time since the start of the scene that a racer has been removed. If you need help finding where in the code to do that let me know

Aether700 commented on Assets/Scripts/PositionTracker.cs:
```diff
@@ -6,16 +6,15 @@ public class PositionTracker : MonoBehaviour
         hasFinishedRace = false;
         RacerManager.AddTracker(this);
-        StartTime = Time.time;
+        scoreManger = GameObject.Find("Race Track").GetComponent<ScoreManager>();
```
Never use GameObject.Find given how slow it is. The ScoreManager is a singleton and therefore you do not need to have a ScoreManager instance to interact with it. 

If you created new functions for the ScoreManager they should be static so that they can be called from anywhere in the code base

Aether700 commented on Assets/Scripts/ScoreManager.cs:
```diff
@@ -56,4 +59,8 @@ public static void AddScore(ScoreData score)
     }
 
     public static List<ScoreData> GetScoreData() { return Instance.racerScores; }
+    public int GetMinScore() { return MinTimeScore; }
```
These methods should be static given that the ScoreManager is a singleton. You should be returning Instance.minTimeScore instead.

Also, to follow the coding style we established the variable names should begin with a non-capitalized letter.

Aether700 reviewed (commented):
There are some issues with how you fetch the first finished racer. I left a comment explaining the details of the problem.

Similarly the ScoreManager is a singleton and therefore you do not need to use GameObject.Find. Again more detail in my comments

WendyFu004 committed:
Followed singleton rule

Now ScoreManager and RacerManager will keep singleton pattern with static functions. When finishRace in the PositionTracker being called  RacerManager.FinishRace() would called before setting the score so the ScoreManager could save the earliest time racer first hits the target.

Aether700 commented on Assets/Scripts/AI/RacerManager.cs:
```diff
@@ -9,6 +9,7 @@ public class RacerManager : MonoBehaviour
     private List<Rigidbody> racerRigidbodies = new List<Rigidbody>();
     private List<PositionTracker> positionTrackers = new List<PositionTracker>();
     private int numFinishedRacers;
+    private bool FirstHit = true;
```
again minor detail but according to our coding style a variable name should start with a non-capitalized letter

Aether700 reviewed (dismissed):
Very good work

WendyFu004 committed:
Change var name to small capitals

Aether700 reviewed (approved):
Good job
